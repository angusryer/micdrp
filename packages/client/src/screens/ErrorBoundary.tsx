import React, { PureComponent, ReactElement } from 'react';
import { Text, View } from 'react-native';
import AppError from '../utilities/errors';

interface ErrorBoundaryProps {
  children: ReactElement;
}

interface State {
  error: AppError | undefined;
}

export default class ErrorBoundary extends PureComponent<ErrorBoundaryProps> {
  state: State = {
    error: undefined
  };

  static getDerivedStateFromError(error: unknown): State {
    return {
      error: new AppError(error)
    };
  }

  /**
   * @param error - unknown error type
   * @param errorInfo - { componentStack }: ErrorInfo
   * */
  async componentDidCatch() {
    // const appError = new AppError(error, { stack: componentStack });
    // Send to log endpoint
  }

  restartApp() {
    // Possibly deal with user credentials before restarting?
    // RNRestart.Restart();
  }

  render() {
    if (this.state.error) {
      return (
        <View>
          <Text>An Error has occurred</Text>
        </View>
      );
    } else {
      return this.props.children;
    }
  }
}
