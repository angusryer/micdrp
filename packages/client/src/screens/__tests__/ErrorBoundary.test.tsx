import { render, screen, waitFor } from '@testing-library/react-native';
import React from 'react';
import { View } from 'react-native';
import ErrorBoundary from '../ErrorBoundary';

describe('ErrorBoundary component', () => {
  const originalConsoleDotError = console.error;

  beforeAll(() => {
    // Suppress console.error that occurs when the error below throws
    console.error = jest.fn();
  });

  afterAll(() => {
    console.error = originalConsoleDotError;
  });

  const TestComponent = ({ shouldThrow }: { shouldThrow?: boolean }) => {
    if (shouldThrow) {
      throw new Error('error');
    } else {
      return <View testID='child'></View>;
    }
  };
  it('renders children properly when no error is thrown', async () => {
    await waitFor(() =>
      render(
        <ErrorBoundary>
          <TestComponent />
        </ErrorBoundary>
      )
    );
    expect(screen.getByTestId('child')).toBeDefined();
    expect(screen).toMatchSnapshot();
  });

  it('renders when an uncaught error is thrown from within a child component', async () => {
    await waitFor(() =>
      render(
        <ErrorBoundary>
          <TestComponent shouldThrow />
        </ErrorBoundary>
      )
    );
    expect(screen.queryByTestId('child')).toBeNull();
    expect(screen).toMatchSnapshot();
  });
});
