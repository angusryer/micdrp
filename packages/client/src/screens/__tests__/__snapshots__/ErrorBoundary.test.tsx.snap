// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary component renders children properly when no error is thrown 1`] = `
<View
  testID="child"
/>
`;

exports[`ErrorBoundary component renders when an uncaught error is thrown from within a child component 1`] = `
<View>
  <Text>
    An Error has occurred
  </Text>
</View>
`;
